Documentation for 0.6 (#884)
AbsaOSS/spline: tmp-0.6/spark-agent0.6.MD

# Spline Spark Agent 0.6

Spark Agent goals are to:
1) capture lineages of executed Spark jobs
2) convert them to Spline lineage format
3) send them to Spline Geatway (using Producer API)

To do this the agent needs to have a listener registered in Spark and also to know the connection to the gateway.

## Spark / Scala version compatibility matrix

|            | Scala 2.11                   | Scala 2.12 |
|------------|:----------------------------:|:----------:|
|**Spark 2.2** | (no SQL; no codeless init) | &mdash;    |
|**Spark 2.3** | (no Delta support)         | &mdash;    |
|**Spark 2.4** | Yes                        | Yes        |
|**Spark 3.0** | &mdash;                    | Yes        |

## Artifacts
- `agent-core_{{SPARK_VERSION}}` is a classic maven library that you can use with any compatible Spark version. This is required if you want to have acces to agent classes or use **programatic initialization**. 

> - `spark-{{SCALA_VERSION}}-spline-agent-bundle_{{SPARK_VERSION}}` is a fat jar containing all dependencies inside omitting the ones already provided by Spark. You can use the bundle in Spark via `--packages` flag. If you also activate the agent via **codeless initialization** your own code doesn't have to have any dependency on the agent it will be used directly by Spark.

Reviewer: Looks like you confused Scala version with Spark version. Can you check the text and fix where needed? (I didn't seek for all occurrences)

Author: I am thinking about removing the Scala version completely from the name. The Scala version at the end is a standard used by everyone. Specifically mentioning it here can make it more confusing.

## Initialization
Initialization is a process of creating agents objects and registering a listener in Spark that will call the agent after each job execution.
### Programatic Initialization

Initialization done in code via calling `enableLineageTracking()` function.
```Scala
// given a Spark session ...
val sparkSession: SparkSession = ???

// ... enable data lineage tracking with Spline
import za.co.absa.spline.harvester.SparkLineageInitializer._
sparkSession.enableLineageTracking()

// ... then run some Dataset computations as usual.
// The lineage will be captured and sent to the configured Spline Producer endpoint.
```

### Codeless Initialization
Initialization using Spark's `spark.sql.queryExecutionListeners` config parameter. This will work with both `spark-submit` and `pyspark`.

```bash
pyspark \
  --packages za.co.absa.spline.agent.spark:spark-2.4-spline-agent-bundle_2.12:0.6.0 \
  --conf "spark.sql.queryExecutionListeners=za.co.absa.spline.harvester.listener.SplineQueryExecutionListener" \
  --conf "spline.lineageDispatcher.http.producer.url=http://localhost:9090/producer"
```

## Configuration

> Agent is looking for configuration properties in those sources:

Reviewer: ```suggestion
The agent looks for configuration properties in the following sources:
```

- Hadoop configuration (core-site.xml)
- Spark configuration
- JVM system properties
- `spline.properties` file on classpath

> File in agent `spline.default.properties` contains default values for properties as well as some additional properties docuemntation.

Reviewer: ```suggestion
The file `spline.default.properties` (in the agent source code) contains default values for properties as well as some additional documentation.
```

It's good idea to look in the file to see what properties are available.

### Properties

`spline.mode`
- `DISABLED` Lineage tracking is completely disabled and Spline is unhooked from Spark.

> - `REQUIRED` If Spline fails to initialize itself (e.g. wrong configuration, no db connection etc) the Spark application aborts with an error.

Reviewer: ```suggestion
- `REQUIRED` If Spline fails to initialize itself (e.g. wrong configuration, no db connection etc) the Spark application aborts with an error. (**Note**: that only concerns Spline initialization routine. If the error happens during lineage capturing, or in the Spline dispatcher, then the target Spark job will have already been finished by that time, and the resulted data will have been persisted, regardless of the `spline.mode` settings. The Spline agent doesn't do any automated rollbacks).
```

- `BEST_EFFORT` (default) Spline will try to initialize itself, but if fails it switches to DISABLED mode allowing the Spark application to proceed normally without Lineage tracking.

`spline.lineageDispatcher` 
- Name of lineage dispatcher in use. More info in **Lineage Dispatchers** chapter.

`spline.IWDStrategy.default.onMissingMetrics` - behaviour when info whether the data was actually written is missing
- `CAPTURE_LINEAGE`
- `IGNORE_LINEAGE` (default)

`spline.postprocessingFilter.classNames`
 - Optional comma separated list of user lineage filters. More info in **Post Processing Lineage Filters** chapter.

## Lineage Dispatchers

> Lineage Dispatchers is a class that is usually (but not always) responsible for sending the converted lineage to gateway. By default `HttpLineageDispatcher` is used.

Reviewer: ```suggestion
The `LineageDispatcher` trait is responsible for sending out the captured lineage information. By default, the `HttpLineageDispatcher` is used, that sends the lineage data to the Spline REST endpoint (see Spline Producer API).
```

To find out which parameters you can use look into `spline.default.properties`. For kafka I would have to define at least these two properties:
```
spline.lineageDispatcher.kafka.topic=foo
spline.lineageDispatcher.kafka.producer.bootstrap.servers=localhost:9092
```

### Creating your own dispatcher
There is also possibility to create your own dispatcher. It must implement `LineageDispatcher` trait and have a constructor with a single parameter of type `org.apache.commons.configuration.Configuration`. To use it you must define name and class and also all other parameters you need. For example:
```
spline.lineageDispatcher=my-dispatcher
spline.lineageDispatcher.my-dispatcher.className=org.example.spline.MyDispatcherImpl
spline.lineageDispatcher.my-dispatcher.prop1=value1
spline.lineageDispatcher.my-dispatcher.prop2=value2
```

## Post Processing Lineage Filters
Filters can be used to enrich the lineage with your own custom data or to remove unwanted data like passwords. All filters are aplied after the Spark plan is converted to Spline DTOs, but before the dispatcher is called.

> The filter must implement `LineageFilter` class and have a constructor with a single parameter of type `org.apache.commons.configuration.Configuration`. For an example you can take a look at `DataSourceUriLineageFilter`. You can use the Configuration instance to acces configuration params and make the filter configurable that way. `LineageFilter`'s methods will provide you not only with lineage DTOs, but also wtih `HarvestingContext` object containing the original Spark plan.
> 
> To use the filter you must set it via a property. For example:
> ```
> spline.postprocessingFilter.classNames=za.co.absa.spline.harvester.postprocessing.DataSourceUriLineageFilter
> ```
> 

Reviewer: ``````suggestion
The procedure how filters are registered and configured is similar to the `LineageDispatcher` registration and configuration procedure.
A custom filter class must implement `za.co.absa.spline.harvester.postprocessing.PostProcessingFilter` trait and declare a constructor with a single parameter of type `org.apache.commons.configuration.Configuration`.
Then register and configure it like this: 
```properties
spline.postProcessingFilter.className=my-filter
spline.postProcessingFilter.my-filter.className=my.awesome.CustomFilter
spline.postProcessingFilter.my-filter.prop1=value1
spline.postProcessingFilter.my-filter.prop2=value2
```

Use pre-registered `CompositePostProcessingFilter` to chain up multiple filters:
```properties
spline.postProcessingFilter=composite
spline.postProcessingFilter.composite.filters=myFilter1,myFilter2
```

(see `spline.default.properties` for details and examples)
``````
